File: Object-Oriented-Programming-CPP-2016-/hw8/BigramDyn.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.Vector;
import java.util.InputMismatchException;
import java.lang.Exception;
/**
 *
 * @author aktenburakk
 */
@SuppressWarnings("unchecked")
public class BigramDyn<T> implements Bigram<T> {

    private int dataType;//thpe of the data in file.
    private int size = 0;//the size of the array.
    private Scanner input ;
    private int max = 0;//holds maximum occurences
    private myPair<Integer ,myPair<T ,T> >[] dynMap;//holds pairs ant their occurences.

    private void addNewElement(T element1 , T element2){
    	//this function make the givin two parameter pair then add the pair to array.

        //temporary pair to be added to array
        myPair<T ,T> temp = new myPair<T , T>(element1 , element2);
        if(size >= dynMap.length)
            reload();
        
        dynMap[size].second = temp;
        dynMap[size].first=new Integer(0);
        
        size++;//update the size
    }
    
    private void reload(){
    	//this function make the capacity of the array greater .
    	//+50
        myPair<Integer ,myPair<T ,T> >[] temp = new myPair[dynMap.length + 50];
        
        for (int i = 0; i < size; i++) 
            temp[i] = dynMap[i];
        
        dynMap = temp;
    }
    private void sort(){//this function sort the array according to ther Integer part.

        for (int i = 0 ; i < size; ++i) {
            for (int j = 0; j < size ; ++j) {
                if(dynMap[j].first < dynMap[i].first ){
                    myPair<Integer ,myPair<T ,T> > temp = dynMap[j];
                    dynMap[j] = dynMap[i];
                    dynMap[i] = temp;
                }
            }
        }    
    }
    
    private void setter(){
    	//this function set the occurences and max occurences
    	//then call tje sort function.

        for(int i = 0 ; i < size ; ++i){
            int count=numOfGrams(dynMap[i].second.first,dynMap[i].second.second);
            dynMap[i].first=count;
            if(count > max)
				max = count;         
        }
        sort();
    }

    public BigramDyn(int data){
    	//this function set dataType and take place from memory for dynMap.
        dataType = data;
        dynMap = new myPair[10];
        for(int i  = 0 ; i < dynMap.length ; ++i)
            dynMap[i] = new myPair<Integer ,myPair<T ,T>>(null , null); 
    }
    
    public void readFile(String fileName) throws Exception {
        //This function reads file and fill the map.
    	//if any problem happend then throw exception.
    	// for example , empty file , bad data ...

        Vector<T> data = new Vector<T> ();//holds the data one by one.

        //dataType is undefined if throw exception here.
        if(dataType <= 0 || dataType > 3){
        	System.err.println("Undefined data thype!!");
        	throw new Exception();
        }

        //reading File/////
        try {
            input = new Scanner(new File(fileName));

             //if file doesn't contain anything.
            if (!input.hasNext()){
        		System.err.println("The file is empty!");
            	throw new Exception();
        	}

        	T next = null ;

        	//this loop read data from the file according to dataType.
        	//the add the data to vector that name is data.
        	while(input.hasNext()){
	        	
		        if(dataType == 1)     
		            (next) = (T)(Object)(input.nextInt());
		        else if (dataType == 2)
	                (next) =  (T)input.next();
		        else if(dataType == 3)
	                (next) = (T)(Object)(input.nextDouble());
	    		
	            data.addElement(next);
			}
			    for(int i = 0 ; i < data.size() - 1 ; ++i)
            		addNewElement(data.get(i) , data.get(i + 1));
        		setter();
        }
        catch (FileNotFoundException ex) {
        	System.err.println("The file is not exist!");
            throw new Exception();
        }
        catch(InputMismatchException e){
	    	System.err.println("Bad Data!!");
	    	throw new Exception();
        }finally{
        	input.close();
        }   
    }

    public int numGrams() {//return size of dynMap.
        return size;
    }

    public int numOfGrams(T first, T second) {//returns the occurences of the given pair.
        int total = 0; //holds occurences

        for(int i = 0 ; i < size ; ++i)
            if(dynMap[i].second.first.equals(first) && dynMap[i].second.second.equals(second))
                total++;
        return total;      
    }
    public String toString(){

    	StringBuilder result = new StringBuilder();//holds the output to print out.
    	Vector<myPair<T ,T>> keys = new Vector<myPair<T ,T>>();//holds the keys

    	for(int i = max ; i >= 0 ; --i){
    		for(int j = 0 ; j < numGrams(); ++j){
    			if(dynMap[j].first == i && !keys.contains(dynMap[j].second)){
    				result.append(dynMap[j].second + "->" + dynMap[j].first + "\n");
    				keys.addElement(dynMap[j].second);
    			}
    		}
    	}
        return result.toString();	
    }
    
}

File: Object-Oriented-Programming-CPP-2016-/hw7/BigramDyn.cpp
/* 
 * File:   BigramDyn.cpp
 * Author: aktenburakk
 * 
 * Created on 10 December 2016, 11:45
 */

#include "BigramDyn.h"
#include "Bigram.h"
#include <iostream>
#include <vector>
#include <string>
using namespace std;

template <class T>
BigramDyn<T>::BigramDyn() : Bigram<T>() , size(0) , capacity(10) ,dataType(2){
    //no parameter constructor sets size to 0 and capacity to 10.
    //then take place from memory for dynData;
    //and sets the dataType to 2 for string
    dynData = new pair<int ,pair<T ,T> >[capacity];
}

template <class T>
BigramDyn<T>::BigramDyn(int type) :
					Bigram<T>(),size(0),capacity(10),dataType(type){
 	//This constructor does same thing like no parameter const.
	//and sets the variable dataType.
    dynData = new pair<int ,pair<T ,T> >[capacity];
}

template <class T>
BigramDyn<T>::BigramDyn(const BigramDyn<T>& other) 
           :size(other.size),capacity(other.capacity),dataType(other.dataType){
    
    //copy constructor
    dynData = new pair<int ,pair<T ,T> >[capacity];
    for(int i = 0 ; i < size ; i++)
        dynData[i] = other.dynData[i];   
}

template <class T>
BigramDyn<T>& BigramDyn<T>:: operator =(const BigramDyn<T>& other){
    
    //overloading assignment operator.
    if(size != other.size && capacity != other.capacity){
        delete [] dynData;
		capacity = other.capacity;
		size = other.size;
		dynData = new pair<int ,pair<T ,T> >[capacity];
    }
    for(int i = 0; i < size ; ++i)
		dynData[i] = other.dynData[i];
			
    return *this;
}

template <class T>
void BigramDyn<T>::readFile(string fileName) throw(Error){

    ifstream inp;
    
    vector<T> datas; //keep all datas one by one.

    inp.open(fileName);//file is opening
    if(inp.fail()){//if file is not exist then throwing Error object
    	cerr<<"No such a file!!"<<endl;
        throw Error();
    }
    
    T data;

    //reads the file until end of file
    while (!inp.eof() && inp >> data)
        datas.push_back(data);
    
    if(!inp.eof()){
        cerr<<"Bad Data!"<<endl;
        inp.close();
        throw Error();
    }
    inp.close();

    if(datas.size() == 0){
        cerr<<"Empty File!!!"<<endl;
        throw Error();
    }

    //adding elements to the dynamic data.
    for(int i = 0 ; i < datas.size() - 1 ; ++i)
    	addElement(datas[i] , datas[i + 1]);

    //this function sets some private variables.
    // more information is in the function implemantation.
    setter();
}

template <class T>
void BigramDyn<T>:: addElement(T element1 , T element2){

	//this function firstly control the capacity if capacity isn't enough
	//then take more place from memory.
    if(size >= capacity)
        reload();
    //then here adds the element.
    (dynData[size]).second.first = element1;
    (dynData[size]).second.second = element2;
    (dynData[size]).first = 0;

    //update size.
    size++;
}
template <class T>
void BigramDyn<T> :: reload(){

	//when capcity of the dynData isn't enough , this function is works.
    capacity += 50;
    pair<int ,pair<T, T> >* newArray = new pair<int ,pair<T, T> >[capacity];
    
    for (int i = 0; i < size; i++) 
        newArray[i] = dynData[i];

    delete [] dynData;
    
    dynData = newArray;
}

template <class T>
void BigramDyn<T>::setter(){

	pair<T,T> temp=dynData[0].second;//holds the pair that has most occurence.
	int max = 0 , count = 1;

	for (int i = 0; i < size; ++i){

		count = numOfGrams(dynData[i].second.first , dynData[i].second.second);
		//sets the ith first value of dynData with its occurence.
		(dynData[i]).first = count;

		if (count > max){
			max = count;
			temp = dynData[i].second;		
		}
		count = 1;
	}
	maxPair = temp;
	numberOfMaxGrams = max;//the number of maximum bigrams in dynData.

	sort();//then sorts the dynData according to pair's occurences.
}

template <class T>
void BigramDyn<T>:: sort(){
	

	//this function sorts the dynData according to
	// pair's occurences from greater to smaller.
    for (int i = 0 ; i < size; ++i) {
        for (int j = 0; j < size ; ++j) {
            if(dynData[j].first < dynData[i].first ){
                pair<int ,pair<T ,T> > temp = dynData[j];
                dynData[j] = dynData[i];
                dynData[i] = temp;
            }
        }
    }
}

template <class T>
void BigramDyn<T>:: print()const{

	vector<pair<T,T> > vec;
	bool flag;

	for (int i = 0; i < size ; ++i)
	{
		flag = false;
		//this loop is for not printing out same value on the screen
		for(pair<T,T> x : vec)
			if(x == dynData[i].second)
				flag = true;				
		//then printing out on the screen the values
		if(flag == false){
			cout <<"("<<dynData[i].second.first<<"-";
			cout <<dynData[i].second.second<<") =>"<<dynData[i].first;
			if(i != size - 1)
				cout << endl;
			vec.push_back(dynData[i].second);
		}
	}
}

template <class T>
pair<T ,T> BigramDyn<T>:: maxGrams(){ return maxPair; }

template <class T>
int BigramDyn<T>:: numGrams()const{ return size;}

template <class T>
int BigramDyn<T>:: numOfGrams(T first , T second)const {
    
    int total = 0; //holds occurences
    
    for(int i = 0 ; i < size ; ++i)
        if(dynData[i].second.first==first && dynData[i].second.second==second)
            total++;
    
    return total; 
}

template <class T>
BigramDyn<T>::~BigramDyn() { //free the memory.
    delete [] dynData;
}

